rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+# RFC #8025 - 2021-08-12 - Internal Tracing
+
+This RFC discusses an integration with the Datadog APM product for Vector's
+internal tracing data, with the goal of enabling faster and more accurate
+diagnosis of performance issues in deployed environments.
+
+## Context
+
+- Existing experimental implementation ([#7929](https://github.com/timberio/vector/pull/7929))
+- Internal instrumentation epic ([#3900](https://github.com/timberio/vector/issues/3900))
+
+## Scope
+
+### In scope
+
+- Instrumenting components with tracing spans
+- Sending internal tracing spans to Datadog APM via the Datadog Agent
+- An opt-in system for internal soak tests and experimental use in customer environments
+
+### Out of scope
+
+- Sending internal traces anywhere other than Datadog APM
+- Sending internal traces directly to Datadog without a local agent
+- An always-on tracing subsystem that is enabled by default in all customer deployments
+- General support for sending tracing data through Vector (i.e. trace data type, sources and sinks)
+
+## Pain
+
+- Debugging Vector performance in users environments requires a lot of back and forth
+- The data available for debugging Vector is relatively high level and doesn't always map clearly to specific components and operations
+
+## Proposal
+
+### User Experience
+
+Vector will gain a new `--enable-datadog-tracing` flag that users can enable at
+start time. When enabled, Vector will forward internal tracing spans to a local
+Datadog Agent with the APM feature enabled. A valid API key will be required to
+start the agent, but no additional configuration of Vector itself is required.
+
+### Implementation
+
+#### Overview
+
+Vector already has some very basic trace-style instrumentation via the
+[`tracing`](https://docs.rs/tracing/) crate that we currently use primary for
+logging. This RFC proposes that we expand that instrumentation to include more
+granular spans around the existing events (i.e. `info!` and friends, not Vector
+events).
+
+With those additional spans in place, we can rely on `tracing` subscribers from
+existing libraries to forward those spans to a local Datadog agent.
+
+#### Instrumentation
+
+How we instrument Vector components with spans should be driven by the kinds of
+questions we'll be looking to answer with the resulting data. In particular, the
+focus should be on the types of questions that are difficult to answer with our
+current tools (metrics, logs, and sometimes Datadog's native profiler).
+
+A typical performance investigation starts at a very high level and works its
+way down into specifics. The first step tends to be identifying which component
+is the bottleneck. This is often done with some intuition and/or experimentation
+(e.g. "Lua tends to be slower, try dropping it and see what happens"), and the
+new utilization metrics are designed to help with exactly this.
```

Comment:
I agree that this is the loop. I'd like to see more discussion of how adding in tracing would shorten this loop. The implication is that we'd see -- from customer environment? -- that such and such internal piece of a component is slow and be able to focus our efforts on that. If so, this gets into the same brand of issue that CPU profile does, how to distinguish a time hog from an actual bottleneck. Getting tracing into customer environments is a step beyond just getting into soak environments, so perhaps motivation from one of the [investigation tickets](https://github.com/timberio/vector/issues/8263) would valuable. Questions I have are, how would tracing as suggested here have shortened the work in the linked investigation, especially compared to the results we got from CPU profiling / metrics.